lightning/src/ln/onion_message.rs

```diff
@@ -0,0 +1,805 @@
+	// [`EncryptedTlvs`] for the hop when constructing the onion packet for sending.
+	//
+	// [`EncryptedTlvs`]: EncryptedTlvs
+	pub encrypted_payload: Vec<u8>,
+}
+
+/// Onion messages can be sent and received to blinded routes, which serve to hide the identity of
+/// the recipient.
+pub struct BlindedRoute {
+	/// To send to a blinded route, the sender first finds a route to the unblinded
+	/// `introduction_node_id`, which can unblind its [`encrypted_payload`] to find out the onion
+	/// message's next hop and forward it along.
+	///
+	/// [`encrypted_payload`]: BlindedHop::encrypted_payload
+	pub introduction_node_id: PublicKey,
+	/// Creators of blinded routes supply the introduction node id's `blinding_point`, which the
+	/// introduction node will use in decrypting its [`encrypted_payload`] to forward the onion
+	/// message.
+	///
+	/// [`encrypted_payload`]: BlindedHop::encrypted_payload
+	pub blinding_point: PublicKey,
+	/// The hops composing the blinded route.
+	pub blinded_hops: Vec<BlindedHop>,
+}
+
+impl BlindedRoute {
+	/// Create a blinded route to be forwarded along `node_pks`. The last node pubkey in `node_pks`
+	/// will be the destination node.
+	pub fn new<Signer: Sign, K: Deref>(node_pks: Vec<PublicKey>, keys_manager: &K) -> Result<Self, ()>
+		where K::Target: KeysInterface<Signer = Signer>,
+	{
+		if node_pks.len() <= 1 { return Err(()) }
+		let secp_ctx = Secp256k1::new();
+		let blinding_secret_bytes = keys_manager.get_secure_random_bytes();
+		let blinding_secret = SecretKey::from_slice(&blinding_secret_bytes[..]).expect("RNG is busted");
+		let (mut encrypted_data_keys, mut blinded_node_pks) = construct_blinded_route_keys(&secp_ctx, &node_pks, &blinding_secret).map_err(|_| ())?;
+		let mut blinded_hops = Vec::with_capacity(node_pks.len());
+		debug_assert_eq!(encrypted_data_keys.len(), blinded_node_pks.len());
+		let mut enc_tlvs_keys = encrypted_data_keys.drain(..);
+		let mut blinded_pks = blinded_node_pks.drain(..);
+
+		for pk in node_pks.iter().skip(1) {
+			let encrypted_tlvs = ControlTlvs::Forward {
+				next_node_id: pk.clone(),
+				next_blinding_override: None,
+			};
+			blinded_hops.push(BlindedHop {
+				blinded_node_id: blinded_pks.next().unwrap(),
+				encrypted_payload: Self::encrypt_payload(encrypted_tlvs, enc_tlvs_keys.next().unwrap()),
+			});
+		}
+
+		// Add the recipient final payload.
+		let encrypted_tlvs = ControlTlvs::Receive { path_id: Some([42; 32]) };
+		blinded_hops.push(BlindedHop {
+			blinded_node_id: blinded_pks.next().unwrap(),
+			encrypted_payload: Self::encrypt_payload(encrypted_tlvs, enc_tlvs_keys.next().unwrap()),
+		});
+
+		Ok(BlindedRoute {
+			introduction_node_id: node_pks[0].clone(),
+			blinding_point: PublicKey::from_secret_key(&secp_ctx, &blinding_secret),
+			blinded_hops,
+		})
+	}
+
+	fn encrypt_payload(payload: ControlTlvs, encrypted_tlvs_ss: SharedSecret) -> Vec<u8> {
+		let (rho, _) = onion_utils::gen_rho_mu_from_shared_secret(encrypted_tlvs_ss.as_ref());
```

Comment:
Actually, I can just update `construct_keys_callback` to only return the `rho` instead of the `SharedSecret`, if that works (since the ss itself is never used iiuc, only the rho)